File: app/models/assessment/ExamFeedbackConfig.java
// SPDX-FileCopyrightText: 2024 The members of the EXAM Consortium
//
// SPDX-License-Identifier: EUPL-1.2

package models.assessment;

import com.fasterxml.jackson.annotation.JsonBackReference;
import com.fasterxml.jackson.databind.annotation.JsonSerialize;
import io.ebean.annotation.EnumValue;
import jakarta.persistence.Entity;
import jakarta.persistence.OneToOne;
import jakarta.persistence.Temporal;
import jakarta.persistence.TemporalType;
import models.base.GeneratedIdentityModel;
import models.exam.Exam;
import org.apache.commons.lang3.builder.EqualsBuilder;
import org.apache.commons.lang3.builder.HashCodeBuilder;
import org.joda.time.DateTime;
import org.springframework.beans.BeanUtils;
import services.datetime.DateTimeAdapter;

@Entity
public class ExamFeedbackConfig extends GeneratedIdentityModel {

    public enum ReleaseType {
        @EnumValue("1")
        ONCE_LOCKED,
        @EnumValue("2")
        GIVEN_DATE,
    }

    private ReleaseType releaseType;

    @OneToOne
    @JsonBackReference
    private Exam exam;

    @Temporal(TemporalType.TIMESTAMP)
    @JsonSerialize(using = DateTimeAdapter.class)
    private DateTime releaseDate;

    public ReleaseType getReleaseType() {
        return releaseType;
    }

    public void setReleaseType(ReleaseType releaseType) {
        this.releaseType = releaseType;
    }

    public Exam getExam() {
        return exam;
    }

    public void setExam(Exam exam) {
        this.exam = exam;
    }

    public DateTime getReleaseDate() {
        return releaseDate;
    }

    public void setReleaseDate(DateTime releaseDate) {
        this.releaseDate = releaseDate;
    }

    public ExamFeedbackConfig copy() {
        ExamFeedbackConfig clone = new ExamFeedbackConfig();
        BeanUtils.copyProperties(this, clone, "id", "exam");
        return clone;
    }

    @Override
    public boolean equals(Object other) {
        if (this == other) return true;
        if (!(other instanceof ExamFeedbackConfig otherConfig)) {
            return false;
        }
        return new EqualsBuilder().append(id, otherConfig.id).build();
    }

    @Override
    public int hashCode() {
        return new HashCodeBuilder().append(id).build();
    }
}

File: app/models/assessment/ExamFeedbackConfig.scala
// SPDX-FileCopyrightText: 2024 The members of the EXAM Consortium
//
// SPDX-License-Identifier: EUPL-1.2

package models.assessment

import com.fasterxml.jackson.annotation.JsonBackReference
import jakarta.persistence.*
import models.base.GeneratedIdentityModel
import models.exam.Exam
import org.joda.time.DateTime
import services.datetime.JsonDateTime

import scala.compiletime.uninitialized

@Entity
class ExamFeedbackConfig extends GeneratedIdentityModel:

  @OneToOne
  @JsonBackReference
  var exam: Exam = uninitialized

  @Temporal(TemporalType.TIMESTAMP)
  @JsonDateTime
  var releaseDate: DateTime = uninitialized

  var releaseType: ExamFeedbackReleaseType = uninitialized

  def copy(): ExamFeedbackConfig =
    val c = new ExamFeedbackConfig
    c.releaseType = releaseType
    c.releaseDate = releaseDate
    c

  override def equals(o: Any): Boolean = o match
    case e: ExamFeedbackConfig => this.id == e.id
    case _                     => false

  override def hashCode: Int = id.toInt

// Java enum aliases: Scala code that references AutoEvaluationConfig.ReleaseType (e.g. in
// pattern matches or service logic) needs a stable path to the Java enum class and its type.
// Without these aliases, callers would have to import AutoEvaluationReleaseType directly,
// which breaks the encapsulation pattern used across the model companion objects.
object ExamFeedbackConfig:
  type ReleaseType = ExamFeedbackReleaseType
  val ReleaseType: Class[ExamFeedbackReleaseType] = classOf[ExamFeedbackReleaseType]
